src/Middleware/InitializeTenancyByRequestData.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Stancl\Tenancy\Middleware;
+
+use Closure;
+use Stancl\Tenancy\Exceptions\TenantCouldNotBeIdentifiedException;
+
+class InitializeTenancyByRequestData
+{
+    /** @var callable */
+    protected $onFail;
+
+    public function __construct(callable $onFail = null)
+    {
+        $this->onFail = $onFail ?? function ($e) {
+            throw $e;
+        };
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  \Illuminate\Http\Request  $request
+     * @param  \Closure  $next
+     * @return mixed
+     */
+    public function handle($request, Closure $next)
+    {
+        if (request()->method() !== 'OPTIONS') {
+            try {
+                $this->parseTenant();
+            } catch (TenantCouldNotBeIdentifiedException $e) {
+                ($this->onFail)($e);
+            }
+        }
+
+        return $next($request);
+    }
+
+    protected function parseTenant()
```

Comment:
Perhaps we could accept `Request $request` here from the `handle()` method and use that instead of the `request()` helper?

Reply:
Sure, any negatives to using the request helper?

Reply:
It's unnecessarily calling the service container when you can just pass one variable